src/bartiq/symbolics/ast_parser.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2024 PsiQuantum, Corp.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""AST-based expression parser.
```

Comment:
Right now this comment mostly addresses how this module is different from the old one.
However, it would be good to also explain how it works, as we'll remove the legacy one some day.

Reply:
I improved the docstring, WDYT?

Reply:
👍